### PR TITLE
Prefer calloc over malloc + memset where possible

### PR DESCRIPTION
**Pull request by AtariDreams**
This offloads the memset work to the OS.

### CODE REVIEW

File: pcap-bpf.c [hunk 1]
```diff
@@ -2611,18 +2611,13 @@ pcap_activate_bpf(pcap_t *p)
...
 		pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
 		    errno, "malloc");
 		status = PCAP_ERROR;
 		goto bad;
 	}
-#ifdef _AIX
-	/* For some strange reason this seems to prevent the EFAULT
-	 * problems we have experienced from AIX BPF. */
-	memset(p->buffer, 0x0, p->bufsize);
-#endif
```
Comment (infrastation) [inline pcap-bpf.c:2625]: It would be better to save this comment even after the zeroing is no longer conditional.

File: pcap-dag.c [hunk 2]
```diff
@@ -1274,9 +1274,9 @@ dag_get_datalink(pcap_t *p)
 	int index=0, dlt_index=0;
 	uint8_t types[255];
 
-	memset(types, 0, 255);
+	memset(types, 0, sizeof(types));
```
Comment (infrastation) [inline pcap-dag.c:1277]: `uint8_t types[255] = { 0 };` would fit the problem space better.

File: pcap-netfilter-linux.c [hunk 3]
```diff
@@ -612,7 +612,7 @@ netfilter_activate(pcap_t* handle)
 
 	if (type == NFLOG) {
 		handle->linktype = DLT_NFLOG;
-		handle->dlt_list = (u_int *) malloc(sizeof(u_int) * 2);
+		handle->dlt_list = (u_int *) malloc(2 * sizeof(u_int));
 		if (handle->dlt_list != NULL) {
```
Comment (infrastation) [inline pcap-netfilter-linux.c:616]: I wonder if this should fail with an error as is done below. In any case, this change is out of scope of this commit.

File: pcap.c [hunk 4]
```diff
@@ -551,8 +551,8 @@ pcap_list_tstamp_types(pcap_t *p, int **tstamp_typesp)
 		**tstamp_typesp = PCAP_TSTAMP_HOST;
 		return (1);
 	} else {
-		*tstamp_typesp = (int*)calloc(sizeof(**tstamp_typesp),
-		    p->tstamp_type_count);
+		*tstamp_typesp = (int*)malloc(p->tstamp_type_count *
+			sizeof(**tstamp_typesp));
```
Comment (infrastation) [inline pcap.c:555]: This is the exact reverse of the point of this commit. Removing this initialization has potential to introduce a bug.

File: pcap.c [hunk 5]
```diff
@@ -3016,14 +3016,14 @@ pcap_list_datalinks(pcap_t *p, int **dlt_buffer)
 		**dlt_buffer = p->linktype;
 		return (1);
 	} else {
-		*dlt_buffer = (int*)calloc(sizeof(**dlt_buffer), p->dlt_count);
+		*dlt_buffer = (int*)malloc(p->dlt_count * sizeof(**dlt_buffer));
```
Comment (infrastation) [inline pcap.c:3019]: Same as above, this change would introduce an unpleasant surprise.

File: gencode.c [hunk 6]
```diff
@@ -632,12 +632,11 @@ newchunk_nolongjmp(compiler_state_t *cstate, size_t n)
 			return (NULL);
 		}
 		size = CHUNK0SIZE << k;
-		cp->m = (void *)malloc(size);
+		cp->m = (void *)calloc(1, size);
```
Comment (infrastation) [inline gencode.c:635]: Both `malloc()` and `calloc()` return a `void *`, so the type cast is noise.

File: pcap.c (hunk 4 above)
Comment (AtariDreams) [reply pcap.c:555]: I have no idea how that got there.

File: pcap.c (hunk 5 above)
Comment (AtariDreams) [reply pcap.c:3019]: Not really, because memset

File: pcap.c (hunk 4 above)
Comment (AtariDreams) [reply pcap.c:555]: memset takes care of that

**Comment by AtariDreams** [comment]
@infrastation Done!

**Comment by infrastation** [comment]
The most recent revision still replaces `calloc()` with `malloc()` without `memset()` twice: in `pcap_list_tstamp_types()` and in `pcap_list_datalinks()`. Excuse me, but we are talking in circles and these changes do not look safe, therefore closing.
